### PR TITLE
New Field Types and Backwards Compatibility

### PR DESCRIPTION
**Pull request by jonyoder**
* Removes some legacy test code.
* Reorganizes some of the code to avoid large files.
* Introduces support for integer and float data types.
* Adds more unit tests for writing simple objects (arrays, strings, etc.).
* Adds the ability to automatically write a struct field index before the first object.
* Adds the ability to read a struct field index from the RSF file.
* Adds the alternate ability to update a reader with a pre-determined struct field index.
* Adds the `Advance` method and its associated interfaces and implementations. These allow you to:
  * Advance to the end of an element.
  * Advance from the start of an element to a known field.
  * Advance from a known field in an element to another known field in the same element.

I would recommend reviewing generally in this order:
* Review the commit that adds int and float support: https://github.com/rstudio/repository-snapshot-format/commit/aeee427cc331fb0a56eca13bcb77a83de6b01167
* Review the new object writing tests: https://github.com/rstudio/repository-snapshot-format/commit/74ad39b29c29127da6c42ee5823cdd9df5f0b639
* Take a look at the new tests for reading an upgraded object from legacy code. This is at https://github.com/rstudio/repository-snapshot-format/blob/5250e04d324fbc284ea764aec1210baa67a8cb7b/writer_upgrade_test.go, and I would get really familiar with how this works before going further.
* Review the rest of the code, skipping what you've already seen.

### CODE REVIEW

**Review by jmwoliver — APPROVED** [review]
This makes sense after walking through it together! Everything looks good, just a few comments. Overall, I'd say it'd be helpful to have more comments on how everything works and why (i.e. like why we append an empty string `""` to `at` in `AdvanceTo()`, those sorts of things would be nice to have more comments for).

File: reader.go [hunk 1]
```diff
@@ -50,6 +58,31 @@ func (f *rsfReader) ReadSizeField(r io.Reader) (int, error) {
 	return int(sz), nil
 }
 
+func (f *rsfReader) ReadInt64Field(r io.Reader) (int64, error) {
+	bs := make([]byte, binary.MaxVarintLen64)
+	i, err := io.ReadFull(r, bs)
+	if err != nil {
+		return 0, err
+	} else if i != binary.MaxVarintLen64 {
+		return 0, fmt.Errorf("unexpected read size %d; expected %d", i, sizeFieldLen)
```
Comment (jmwoliver) [inline reader.go:67]: This comment didn't go through for some reason, but this should be updated too:

```suggestion
		return 0, fmt.Errorf("unexpected read size %d; expected %d", i, binary.MaxVarintLen64)
```

**Comment by jonyoder** [comment]
@jmwoliver I think I updated everything based on your comments. I also made the `ReadIndex` method return the actual index, which will be useful for the RSF provider in PPM.

**Review by jmwoliver — COMMENTED** [review]
Just a few more comments, but feel free to ignore since you already merged!

**Pull request merged**